de.fu_berlin.inf.dpp.core/src/de/fu_berlin/inf/dpp/session/SarosSessionManager.java

```diff
@@ -532,44 +550,105 @@ public void addResourcesToSession(
             return;
         }
 
-        List<IProject> projectsToShare = new ArrayList<IProject>();
+        /** add new shared resources to a combining map for the next negotiation **/
+        synchronized (this) {
+            for (Entry<IProject, List<IResource>> mapEntry : projectResourcesMapping
+                .entrySet()) {
+                final IProject project = mapEntry.getKey();
+                final List<IResource> resourcesList = mapEntry.getValue();
 
-        for (Entry<IProject, List<IResource>> mapEntry : projectResourcesMapping
-            .entrySet()) {
+                /** case: first entry for project **/
+                if (!nextSharedResources.containsKey(project)) {
+                    if (resourcesList == null || resourcesList.isEmpty())
+                        nextSharedResources.put(project, null);
+                    else {
+                        nextSharedResources.put(project,
+                            new ArrayList<IResource>(resourcesList));
+                    }
+                } else {
+                    List<IResource> list = nextSharedResources.get(project);
+
+                    /** case: project is already a completely shared **/
+                    if (list == null || list.isEmpty())
+                        continue;
+                    /** case: update partial project to completely shared **/
+                    if (resourcesList == null || resourcesList.isEmpty())
+                        nextSharedResources.put(project, null);
+                    else
+                        /** case: just add new resources to a queued project **/
+                        list.addAll(resourcesList);
+                }
+            }
 
-            final IProject project = mapEntry.getKey();
-            final List<IResource> resourcesList = mapEntry.getValue();
+            /** leave if a thread is already waiting to do the work **/
+            if (nextSharedResourcesWorker) {
+                return;
+            }
+            /** otherwise become the working thread **/
+            nextSharedResourcesWorker = true;
+        }
 
-            // side effect: non shared projects are always partial -.-
-            if (!currentSession.isCompletelyShared(project)) {
-                String projectID = currentSession.getProjectID(project);
+        /**
+         * leave the synchronized block to allow new resource adding and wait
+         * till all running negotiations are done
+         **/
+        while (currentProjectNegotiations.negotiationsInProgress()) {
+            try {
+                synchronized (nextSharedResourcesWorkerNotify) {
+                    nextSharedResourcesWorkerNotify.wait();
+                }
+            } catch (InterruptedException e) {
```

Comment:
![Bug](https://sonarsource.github.io/sonar-github/bug.png 'Bug') Bug: Either re-interrupt this method or rethrow the "InterruptedException". ([squid:S2142](https://sonarcloud.io/organizations/saros/rules#rule_key=squid%3AS2142))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=saros_complete&pullRequest=272&issues=AWaT0XAZyONdBoB89fGy&open=AWaT0XAZyONdBoB89fGy)

[ISSUE_KEY]: AWaT0XAZyONdBoB89fGy